crypto/x509/x509_vpm.c

```diff
@@ -313,8 +321,50 @@ int X509_VERIFY_PARAM_set1(X509_VERIFY_PARAM *to,
   return ret;
 }
 
-static int int_x509_param_set1(char **pdest, size_t *pdestlen, const char *src,
-                               size_t srclen) {
+static int int_x509_param_set1_email(char **pdest, size_t *pdestlen,
+                                     const char *src, size_t srclen) {
+  void *tmp;
+  if (src != NULL) {
+    // Setting |srclen| to 0 to automatically detect the length of |src| is an
+    // OpenSSL quirk that AWS-LC isn't keen on supporting. However, consumers
+    // often assume OpenSSL semantics from AWS-LC, so it's supported in this
+    // case.
+    if (srclen == 0) {
+      srclen = strlen(src);
+    }
```

Comment:
Should this have an else that ensures there are no internal \0 like int_x509_param_set_hosts? I see the original code in int_x509_param_set1 didn't enforce this for ip address or emails, any idea why they are different?

Reply:
IP addresses are allowed to have NULL bytes due to their nature: https://github.com/aws/aws-lc/blob/8402a6eecfff195793a57b5bf1cf371f1218da3c/crypto/x509/x509_test.cc#L1790-L1791

We could look into enforcing this for emails if we wanted to? `int_x509_param_set_hosts` was a different function since it deals with a Stack instead of a single string like the others.

Reply:
Emails already have this enforced in `X509_VERIFY_PARAM_set1_email`. We also have a test for it.
* https://github.com/aws/aws-lc/blob/15bd84a7a63b44c7a94965a13a9df74b1db25483/crypto/x509/x509_vpm.c#L457
* https://github.com/aws/aws-lc/blob/main/crypto/x509/x509_test.cc#L1749-L1755